### DISCUSSION TITLE
Rendering Markdown files into a single page

### DISCUSSION THREAD

**Discussion by theranger** [statiqdev/Discussions#205] [Q&A]
I am trying to use Statiq to combine multiple *.md input files to a single page HTML file as sections one after another. Something like this came to my mind at first:

```
ArchiveSources: => "*.md"
ArchiveOrderKey: Order
ArchiveOrderDescending: false
IncludeInSitemap: false
---

@foreach (var post in Model.GetChildren()) {
	var image = post.GetString("Image");
	<a id="@post.GetLink().TrimStart('/')" class="@post.GetString("Style")">
		<img src="/img/@image" alt="Image" />
		<span>@Html.Raw(await post.GetContentStringAsync())</span>
	</a>
}
```

This works. However I would like to have every child to be rendered using a custom layout declared in the FrontMatter header section of the every Markdown file. How to use the fully rendered output (Markdown -> HTML -> Razor Template) to render every child element in the sample above?

Have been experimenting with a custom pipeline as well (taking the Archive pipeline as a basis) but it got really messy. I managed to combine the documents, but I could not get the post-process rendering working properly with some crucial bits missing from the code:

```
ProcessModules = new ModuleList {
	new GetPipelineDocuments(Config.FromDocument(doc => doc.Get<ContentType>(WebKeys.ContentType) != ContentType.Asset || doc.MediaTypeEquals(MediaTypes.CSharp))),
	new FilterDocuments(Config.FromDocument(IsArchive)),
	
	new ForEachDocument {
		new ExecuteConfig(Config.FromDocument((archiveDoc, _) => {
			return new ModuleList {
				new ReplaceDocuments(archiveDoc.GetList(WebKeys.ArchivePipelines, new[] { nameof(Content) }).ToArray()),
				new MergeMetadata(Config.FromValue(archiveDoc.Yield())).KeepExisting(),
				new FilterSources(archiveDoc.GetList<string>(WebKeys.ArchiveSources)),
				new OrderDocuments(archiveDoc.GetString(WebKeys.ArchiveOrderKey)),
				templates.GetModule(ContentType.Content, Phase.PostProcess),
				new CombineDocuments(),
			};
		}))			
	}
};

PostProcessModules = new ModuleList {
	new SetDestination("output.html"),
	new ExecuteSwitch(Config.FromDocument(doc => doc.Get<ContentType>(WebKeys.ContentType)))
		.Case(ContentType.Data, templates.GetModule(ContentType.Data, Phase.PostProcess))
		.Case(ContentType.Content, (IModule)new RenderContentPostProcessTemplates(templates))
};
```

Any ideas?

**Comment by TheBrambleShark** [comment]
You should be able to do this without a module simply by using the existing Razor subsystem.

In the frontmatter for each markdown file, define the `layout` key. I use a cshtml file for the layout.

SamplePage.md
```md
---
Title: Sample Page
xref: sample
layout: _SampleLayout
---
Hello World!
```

_SampleLayout.cshtml
```cshtml
@{
    // If you have a _MasterLayout.cshtml file in the root which gives you your overall site layout, specify that here:
    Layout = "_MasterLayout";
}
@* You can have whatever layout you like here. This is representative of the layout used for each markdown file. *@
<div class="row">
  <div class="col">
    <partial name="_tableOfContents" />
  </div>
  <div class="col">
    @RenderBody()
  </div>
  <div class="col">
    <partial name="_sidebar" />
</div>
```

Index.cshtml
```cshtml
@{
    DocumentList<IDocument> children = OutputPages.GetChildrenOf(Document);
}

<div class="document-list">
@foreach (IDocument document in children)
{
    @(await document.GetContentStringAsync());
}
```

Getting the content string may not be exactly right, but it should be fairly close to what you're looking for. Hopefully this can at least set you on the right track.

**Reply by theranger** [reply]
Well, this is exactly what I explained. Declaring a layout does not help, since `GetContentStringAsync()` returns the HTML representation of the Markdown module rendering output, not the full page rendered by Razor module.

Rendering the full Razor layout is done at the post-processing phase where the content is already merged into the final document. The content in this case is pure HTML derived from Markdown, not the full Razor template.

**Comment by theranger** [comment]
I would pose the question this way:

How to use Markdown files to render partial Razor layouts with content and then use some logic (e.g. pipelines) to combine those intermediate partial outputs to one single page?
